### PR TITLE
Improvement: Custom Scoreboard Refactor

### PR DESCRIPTION
**Pull request by j10a1n15**
## What
This Pull Request cleans up and refactors the whole Custom Scoreboard. It switches from enums with displaypair getters to abstract classes of each entry, and many more changes.
Also uhh yes these are a few additions, we are sorry hanni <3.
More changes to improve performance are planned for a later PR.

<details>
<summary>Images</summary>

<details>
<summary>Text alignment</summary>

image
image

</details>

</details>

## Changelog Improvements
+ Made multiple improvements to the Custom Scoreboard. - Empa, j10a1n15
    * Added an option to align the text.
    * Improved overall performance.
    * Added the Party Leader to the Party Element.
    * Separated title and footer alignment.
    * Added a custom alpha footer.

## Changelog Technical Details
+ Added `SizeLimitedCache` and `SizeLimitedSet`. - Empa
+ Removed the deprecated pattern functions inside `StringUtils`. - Empa
+ Refactored Custom Scoreboard. - Empa, j10a1n15
    * Abstracted each Element/Event into separate classes extending `ScoreboardConfigElement` for a more modular design.
    * Made it possible to declare elements in different ways, such as using single strings or lists of strings.
    * Optimized various elements to use widgets or already stored values.
    * Optimized by only checking elements that can be shown on the current island.
    * Optimized unknown lines handling by storing recently checked lines in a `SizeLimitedSet`.

### CODE REVIEW

**Review by hannibal002 — CHANGES_REQUESTED** [review]
so many small files in events and in elements is bad. please merge the objects into multiple bigger files. also give the names abstract names that group them together cleanly, and give all classes a name that contains some hint that this is scoreboard related.

File: src/main/java/at/hannibal2/skyhanni/features/gui/customscoreboard/ChunkedStatsManager.kt [hunk 1]
```diff
@@ -4,78 +4,109 @@ import at.hannibal2.skyhanni.data.MiningAPI
...
+import at.hannibal2.skyhanni.features.gui.customscoreboard.elements.ScoreboardElementGems
+import at.hannibal2.skyhanni.features.gui.customscoreboard.elements.ScoreboardElementHeat
+import at.hannibal2.skyhanni.features.gui.customscoreboard.elements.ScoreboardElementMotes
+import at.hannibal2.skyhanni.features.gui.customscoreboard.elements.ScoreboardElementNorthStars
+import at.hannibal2.skyhanni.features.gui.customscoreboard.elements.ScoreboardElementPurse
 
 private val hideEmptyLines get() = informationFilteringConfig.hideEmptyLines
 
-enum class ChunkedStat(
+enum class ChunkedStatsManager(
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/gui/customscoreboard/ChunkedStatsManager.kt]: This is no manager, this is the enum element. use a different name

File: src/main/java/at/hannibal2/skyhanni/features/gui/customscoreboard/ChunkedStatsManager.kt [hunk 2]
```diff
@@ -4,78 +4,109 @@ import at.hannibal2.skyhanni.data.MiningAPI
...
     ),
     ;
 
     override fun toString() = configLine
 
     companion object {
-        fun getChunkedStats() = chunkedConfig.chunkedStats.mapNotNull {
-            it.takeIf { it.showWhen() }?.displayPair?.invoke()
+
+        private var currentIslandStats = listOf<ChunkedStatsManager>()
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/gui/customscoreboard/ChunkedStatsManager.kt]: move the companion object members into a real manager class

File: src/main/java/at/hannibal2/skyhanni/api/SkillAPI.kt [hunk 3]
```diff
@@ -318,17 +318,18 @@ object SkillAPI {
         }
 
         val existingLevel = getSkillInfo(skillType) ?: SkillInfo()
+        val gained = matcher.group("gained")
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/api/SkillAPI.kt]: unrelated changes like those should really not be done in this pr

File: src/main/java/at/hannibal2/skyhanni/config/enums/OutsideSbFeature.kt [hunk 4]
```diff
@@ -24,5 +24,5 @@ enum class OutsideSbFeature(private val displayName: String) {
 
     override fun toString() = displayName
 
-    fun isSelected() = Minecraft.getMinecraft().thePlayer != null && SkyHanniMod.feature.misc.showOutsideSB.get().contains(this)
+    fun isSelected() = Minecraft.getMinecraft().thePlayer != null && this in SkyHanniMod.feature.misc.showOutsideSB.get()
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/config/enums/OutsideSbFeature.kt]: again unrelated changes that unnecessary blow up the pr size

File: src/main/java/at/hannibal2/skyhanni/features/gui/customscoreboard/ScoreboardEntry.kt [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+import at.hannibal2.skyhanni.features.gui.customscoreboard.elements.ScoreboardElement
+import at.hannibal2.skyhanni.features.gui.customscoreboard.elements.ScoreboardElementSlayer
+import at.hannibal2.skyhanni.features.gui.customscoreboard.elements.ScoreboardElementSoulflow
+import at.hannibal2.skyhanni.features.gui.customscoreboard.elements.ScoreboardElementTime
+import at.hannibal2.skyhanni.features.gui.customscoreboard.elements.ScoreboardElementTitle
+import at.hannibal2.skyhanni.features.gui.customscoreboard.elements.ScoreboardElementTuning
+import at.hannibal2.skyhanni.features.gui.customscoreboard.elements.ScoreboardElementUnknown
+import at.hannibal2.skyhanni.features.gui.customscoreboard.elements.ScoreboardElementVisiting
+
+enum class ScoreboardEntry(val element: ScoreboardElement) {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/gui/customscoreboard/ScoreboardEntry.kt]: better name than entry here

File: src/main/java/at/hannibal2/skyhanni/config/features/gui/customscoreboard/CustomScoreboardConfig.java [hunk 6]
```diff
@@ -28,16 +28,17 @@ public class CustomScoreboardConfig {
...
     @ConfigEditorDraggableList()
-    public List<ScoreboardElement> scoreboardEntries = new ArrayList<>(ScoreboardElement.defaultOption);
+    public Property<List<ScoreboardEntry>> scoreboardEntries = Property.of(new ArrayList<>(ScoreboardEntry.defaultOption));
 
     @ConfigOption(name = "Reset Appearance", desc = "Reset the appearance of the advanced scoreboard.")
     @ConfigEditorButton(buttonText = "Reset")
     public Runnable reset = () -> {
-        scoreboardEntries.clear();
-        scoreboardEntries.addAll(ScoreboardElement.defaultOption);
+        scoreboardEntries.get().clear();
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/config/features/gui/customscoreboard/CustomScoreboardConfig.java]: in the future this kind of behaviour/logic belongs into kotlin scope

File: src/main/java/at/hannibal2/skyhanni/config/features/gui/customscoreboard/CustomScoreboardConfig.java [hunk 7]
```diff
@@ -56,7 +57,7 @@ public class CustomScoreboardConfig {
     public InformationFilteringConfig informationFiltering = new InformationFilteringConfig();
 
     @Expose
-    @ConfigOption(name = "Unknown Lines warning", desc = "Give a chat warning when unknown lines are found in the scoreboard.")
+    @ConfigOption(name = "Unknown Lines warning", desc = "Give a chat warning when unknown lines are found in the scoreboard.\n§cReporting these in the Discord Server are very important, so we can know what lines are missing.")
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/config/features/gui/customscoreboard/CustomScoreboardConfig.java]: line too long, split line at line breah character

File: src/main/java/at/hannibal2/skyhanni/config/features/gui/customscoreboard/DisplayConfig.java [hunk 8]
```diff
@@ -145,8 +145,12 @@ public String toString() {
...
 
     @Expose
-    @ConfigOption(name = "Cache Scoreboard on Island Switch",
-        desc = "Will stop the Scoreboard from updating while switching islands.\nRemoves the shaking when loading data.")
+    @ConfigOption(name = "Text Alignment", desc = "Will align the text to the left, center or right, while not overriding certain lines, like title or footer.")
+    @ConfigEditorDropdown
+    public RenderUtils.HorizontalAlignment textAlignment = RenderUtils.HorizontalAlignment.LEFT;
+
+    @Expose
+    @ConfigOption(name = "Cache Scoreboard on Island Switch", desc = "Will stop the Scoreboard from updating while switching islands.\nRemoves the shaking when loading data.")
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/config/features/gui/customscoreboard/DisplayConfig.java]: line too long, split

File: src/main/java/at/hannibal2/skyhanni/config/features/gui/customscoreboard/PartyConfig.java [hunk 9]
```diff
@@ -4,21 +4,25 @@
...
     @Expose
     @ConfigOption(name = "Max Party List", desc = "Max number of party members to show in the party list (you are not included).")
     @ConfigEditorSlider(
         minValue = 1,
-        maxValue = 25, // why do I even set it so high
+        maxValue = 25,
         minStep = 1
     )
-    public Property<Integer> maxPartyList = Property.of(4);
+    public Integer maxPartyList = 4;
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/config/features/gui/customscoreboard/PartyConfig.java]: why is this no longer a  observer? this is counter productive to the plan of having the feature react to changes and not periodically udpate everything

File: src/main/java/at/hannibal2/skyhanni/data/HypixelData.kt [hunk 10]
```diff
@@ -485,32 +485,34 @@ object HypixelData {
...
+            newIsland = getIslandType(foundIsland, guesting)
         }
 
         // TODO don't send events when one of the arguments is none, at least when not on sb anymore
-        if (skyBlockIsland != islandType) {
-            IslandChangeEvent(islandType, skyBlockIsland).postAndCatch()
-            if (islandType == IslandType.UNKNOWN) {
+        if (skyBlockIsland != newIsland) {
+            val oldIsland = skyBlockIsland
+            skyBlockIsland = newIsland
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/data/HypixelData.kt:507]: useful change, but also feels a bit out of scope of the pr

File: src/main/java/at/hannibal2/skyhanni/data/MaxwellAPI.kt [hunk 11]
```diff
@@ -135,13 +135,14 @@ object MaxwellAPI {
 
         chatPowerPattern.tryReadPower(message)
         chatPowerUnlockedPattern.tryReadPower(message)
-        tuningAutoAssignedPattern.matchMatcher(event.message) {
+        if (tuningAutoAssignedPattern.matches(event.message)) {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/data/MaxwellAPI.kt]: early return

File: src/main/java/at/hannibal2/skyhanni/data/MaxwellAPI.kt [hunk 12]
```diff
@@ -135,13 +135,14 @@ object MaxwellAPI {
...
             if (tunings.isNullOrEmpty()) return
-            val tuningsInScoreboard = ScoreboardElement.TUNING in CustomScoreboard.config.scoreboardEntries
-            if (tuningsInScoreboard) {
-                ChatUtils.chat(
-                    "Talk to Maxwell and open the Tuning Page again to update the tuning data in scoreboard.",
-                )
+            with(CustomScoreboard.config) {
+                if (enabled.get() && ScoreboardEntry.TUNING in scoreboardEntries.get()) {
+                    ChatUtils.chat(
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/data/MaxwellAPI.kt]: remove unnecessary new lines

File: src/main/java/at/hannibal2/skyhanni/data/MiningAPI.kt [hunk 13]
```diff
@@ -51,6 +51,10 @@ object MiningAPI {
         "cold.reset",
         "§6The warmth of the campfire reduced your §r§b❄ Cold §r§6to §r§a0§r§6!|§c ☠ §r§7You froze to death§r§7\\.",
     )
+    val coldPattern by group.pattern(
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/data/MiningAPI.kt:57]: regex test

File: src/main/java/at/hannibal2/skyhanni/data/MiningAPI.kt [hunk 14]
```diff
@@ -150,11 +154,15 @@ object MiningAPI {
         IslandType.SPIDER_DEN,
     )
 
+    fun inAdvancedMiningIsland() = inAnyIsland(IslandType.DWARVEN_MINES, IslandType.CRYSTAL_HOLLOWS, IslandType.MINESHAFT)
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/data/MiningAPI.kt:160]: while being useful changes, again outside of the scope of the pr

File: src/main/java/at/hannibal2/skyhanni/features/combat/SpidersDenAPI.kt [hunk 15]
```diff
@@ -1,11 +1,23 @@
...
+import at.hannibal2.skyhanni.utils.RegexUtils.anyMatches
+import at.hannibal2.skyhanni.utils.repopatterns.RepoPattern
 
 @SkyHanniModule
 object SpidersDenAPI {
-    fun isAtTopOfNest(): Boolean = ScoreboardData.sidebarLinesFormatted.any { ScoreboardPattern.broodmotherPattern.matches(it) }
+
+    private val repoGroup = RepoPattern.group("combat.spidersden")
+
+    val broodmotherPattern by repoGroup.pattern(
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/combat/SpidersDenAPI.kt:18]: regex test

File: src/main/java/at/hannibal2/skyhanni/features/gui/customscoreboard/CustomScoreboard.kt [hunk 16]
```diff
@@ -222,20 +219,36 @@ object CustomScoreboard {
...
-                        "${element.name.firstLetterUppercase()} - " +
-                            "${element.showWhen.invoke()} - " +
-                            "${element.getVisiblePair().map { it.first }}",
+                        "   ${entry.name.firstLetterUppercase()} - " +
+                            "island: ${entry.element.showIsland()} - " +
+                            "show: ${entry.element.showWhen()} - " +
+                            "${entry.element.getLines().map { it.display }}",
                     )
                 }
+                allUnknownLines.toSet().takeIfNotEmpty()?.let { set ->
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/gui/customscoreboard/CustomScoreboard.kt]: is the `toSet` call necessary here?

**Review by hannibal002 — CHANGES_REQUESTED** [review]
reviewed around 1/3 of the lines. Not tested in game.
Not all changes requested are actually meant to get changed, for some, I'm just curious why this approach is done. lets dicsuss please.

Also ive merged the changelog lines into one line with sub parts

File: src/main/java/at/hannibal2/skyhanni/data/HypixelData.kt (hunk 10 above)
Comment (ItsEmpa) [reply src/main/java/at/hannibal2/skyhanni/data/HypixelData.kt:507]: this change is needed for this pr to work

File: src/main/java/at/hannibal2/skyhanni/config/features/gui/customscoreboard/TitleAndFooterConfig.java [hunk 17]
```diff
@@ -32,10 +30,22 @@ public class TitleAndFooterConfig {
...
+    @ConfigEditorDropdown
+    public RenderUtils.HorizontalAlignment alignFooter = RenderUtils.HorizontalAlignment.LEFT;
+
     @Expose
     @ConfigOption(name = "Custom Footer", desc = "What should be displayed as the footer of the scoreboard." +
-        "\nUse & for colors." +
+        "\nUse && for colors." +
+        "\nUse \"\\n\" for new line.")
+    @ConfigEditorText
+    public String customFooter = "&&ewww.hypixel.net";
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/config/features/gui/customscoreboard/TitleAndFooterConfig.java:43]: need config migration

**Review by hannibal002 — CHANGES_REQUESTED** [review]
Only bug i could found
image

File: src/main/java/at/hannibal2/skyhanni/config/features/gui/customscoreboard/TitleAndFooterConfig.java (hunk 17 above)
Comment (j10a1n15) [reply src/main/java/at/hannibal2/skyhanni/config/features/gui/customscoreboard/TitleAndFooterConfig.java:43]: [migrator was heavily outdated](https://github.com/hannibal002/SkyHanni/pull/2162/commits/1fb98604d89282854c8919efa0b9332c23ee3ef5)

**Review by hannibal002 — CHANGES_REQUESTED** [review]
image

**Review by hannibal002 — APPROVED** [review]
image
